src/app/icd/client/DefaultCheckInDelegate.cpp

```diff
@@ -44,5 +44,44 @@ void DefaultCheckInDelegate::OnCheckInComplete(const ICDClientInfo & clientInfo)
 #endif
 }
 
+ICDRefreshKeyInfo * DefaultCheckInDelegate::OnKeyRefreshNeeded(const ICDClientInfo & clientInfo,
+                                                               const ICDClientStorage * clientStorage)
+{
+    auto icdRefreshKeyInfo = CreateICDRefreshKeyInfo(clientInfo, clientStorage);
+
+    if (icdRefreshKeyInfo == nullptr)
+    {
+        return nullptr;
+    }
+
+    CHIP_ERROR err = chip::Crypto::DRBG_get_bytes(icdRefreshKeyInfo->mNewKey.Bytes(), icdRefreshKeyInfo->mNewKey.Capacity());
+    if (err != CHIP_NO_ERROR)
+    {
+        return nullptr;
+    }
+
+    return icdRefreshKeyInfo;
+}
+
+void DefaultCheckInDelegate::OnRegistrationUpdateComplete(const ICDClientInfo & clientInfo, ICDRefreshKeyInfo * icdRefreshKeyInfo)
+{
+    ChipLogProgress(ICD, "Re-registration with new key successful. New start counter = %d", clientInfo.start_icd_counter);
+    ReleaseICDRefreshKeyInfo(icdRefreshKeyInfo);
+}
+void DefaultCheckInDelegate::OnRegistrationUpdateFailure(ICDRefreshKeyInfo * icdRefreshKeyInfo)
+{
+    ReleaseICDRefreshKeyInfo(icdRefreshKeyInfo);
+}
+
+ICDRefreshKeyInfo * DefaultCheckInDelegate::CreateICDRefreshKeyInfo(const ICDClientInfo & clientInfo,
+                                                                    const ICDClientStorage * clientStorage)
+{
+    return chip::Platform::MakeUnique<ICDRefreshKeyInfo>(this, clientInfo, clientStorage).get();
```

Comment:
Please figure out why this line is wrong, explain it in a response to this comment, and don't ever write this sort of code again.

Reply:
get() returns a temporary object and it is not okay to use that.

Reply:
More precisely, it returns a pointer to a thing that will be deleted when the return value of `MakeUnique` goes out of scope....